### PR TITLE
Docs 101

### PR DESCRIPTION
**Pull request by ljsinclair**
Hi there
This is the rejig of the data types doco. that needs to be reviewed to make sure it's accurate and my rewrites make sense.

Content found under /sql-preview/data-types

The site apparently builds but I can't visually check it because the server is playing up
There are no broken links after running a broken link checker over it (a new one called html-proofer because as I say, the server is playing up and the python link check wasn't firing up properly)

Worst case, we'll end up doing what we did last week, and transfer the new content to a new branch off of main :)

Thanks!

L.

### CODE REVIEW

File: concepts/_posts/2023-01-01-data-modeling-new.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+Need to gather high level information on the databases here.
```
Comment (gthrone25) [inline concepts/_posts/2023-01-01-data-modeling-new.md:2]: This should be in a separate PR. I'm thinking it's here on accident as we already redid this page??

File: concepts/_posts/2023-01-01-featurebase-tables.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+---
+title: FeatureBase tables overview
```
Comment (gthrone25) [inline concepts/_posts/2023-01-01-featurebase-tables.md:2]: This should be in a separate PR. I'm thinking it's here on accident as we already redid this page??

File: sql-preview/data-types/_posts/2022-11-16-data-types-home.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+Constraints are applied when a column is created.
+
+| Constraint | Data type |
+|---|---|
+| MIN, MAX | [INT](/sql-preview/data-types/cloud-int-data-type) |
+| SCALE | [DECIMAL](/sql-preview/data-types/cloud-decimal-data-type) |
+| TIMEUNIT | [TIMESTAMP](/sql-preview/data-types/cloud-timestamp-data-type) |
+| TIMEQUANTUM, TTL | [IDSET](/sql-preview/data-types/cloud-idset-data-type) |
+| TIMEQUANTUM, TTL | [STRINGSET](/sql-preview/data-types/cloud-stringset-data-type) |
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-data-types-home.md:50]: Can we just have two links for the two data types in on row for Timeqauntum?

File: sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+---
+title: BOOL data type
+---
+
+## Syntax
+
+```
+BOOL...
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md:8]: why the three dots?

File: sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+| BOOL | The BOOL type stores simple Boolean (true/false) values and is used for simple query filtering. |
+
+## Additional information
+
+Use this table to determine how integer and string values are represented when added to a Boolean column.
+
+| data | value | bool |
+|---|---|---|
+| integer | 0 | 0 |
+| integer > 0 | 1 |
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md:24]: Missing the bool column should be:

| integer > 0 | 1 | 1 |

File: sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+| string | empty, `0`, `f`, `false` | 0 |
+| string | other string values | 1 |
+
+## Examples
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md:28]: What do you envision here? I don't see this on the other type pages

File: sql-preview/data-types/_posts/2022-11-16-cloud-decimal-data-type.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
+---
+title: Decimal data type
+---
+
+## Syntax
+
+```
+DECIMAL...SCALE...
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-decimal-data-type.md:8]: DECIMAL(<SCALE>) I think is what we want here

File: sql-preview/data-types/_posts/2022-11-16-cloud-decimal-data-type.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+```
+
+## Arguments
+
+| Argument | Description |
+|---|---|
+| DECIMAL | Numeric data type ideally used for decimal numbers where the exact scale is known. Defaults to two (2) |
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-decimal-data-type.md:15]: Axe Defaults to two (2) as it's right below?

File: sql-preview/data-types/_posts/2022-11-16-cloud-id-data-type.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
+---
+title: ID numeric datatype
+---
+
+## Syntax
+
+```
+ID...
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-id-data-type.md:8]: three dots?

File: sql-preview/data-types/_posts/2022-11-16-cloud-idset-data-type.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
+---
+title: IDSET data type
+---
+
+IDSET is a FeatureBase datatype used with `timeQuantum` and `ttl` constraints.
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-idset-data-type.md:5]: If we name constraints at the top of the page for each type, we should add this to decimal, int, etc.

File: sql-preview/data-types/_posts/2022-11-16-cloud-idset-data-type.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+## API syntax
+
+```
+IDSET...
+```
+
+## Arguments
+
+| Argument | Description | Further information |
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-idset-data-type.md:15]: Why the further information column if we have additional information below? It seems this table heading is in some pages (ID, IDSET, etc.) but not others (Decimal)

File: sql-preview/data-types/_posts/2022-11-16-cloud-idset-data-type.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  * grouping by
+  * searching for discrete values
+* Use the INT data type to perform range queries using `<` or `>`
+
+## Example
+
+Track all the store IDs a customer has visited
+
+```
+CODE EXAMPLE
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-idset-data-type.md:32]: Query example? DDL? Data example? I can help tackle this outside of this PR, but we can probably cut it for now

File: sql-preview/data-types/_posts/2022-11-16-cloud-int-data-type.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+---
+
+Used with integers between `-2^63` and `2^63 -1`.
+
+## Syntax
+
+```
+syntax goes here
+
+INT [MIN | MAX]
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-int-data-type.md:12]: I think we want:

INT MIN <min value> | MAX <max value>

File: sql-preview/data-types/_posts/2022-11-16-cloud-int-data-type.md [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+| Argument | Description | Further information |
+|---|---|---|
+| INT | Used for integer data that spans a large range of values intended for aggregate queries |  |
+| MIN | Minimum value constraint defaults to -2^63 |  |
+| MAX | Maximum value constraint defaults to 2^63 -1 |  |
+
+## Additional information
+
+INT is not suitable for queries that
+* group by
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-int-data-type.md:26]: This can be removed

File: sql-preview/data-types/_posts/2022-11-16-cloud-int-data-type.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+* look for discrete values
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-int-data-type.md:27]: @pokeeffe-molecula thoughts on these three. The key is if your integer is low cardinality and you are doing group bys and discrete lookups, IDs might be better suited. That being said, you can group by and search for discrete ints still

File: sql-preview/data-types/_posts/2022-11-16-cloud-string-data-type.md [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+```
+STRING...
+```
+
+## Arguments
+
+| Argument | Description | Further information |
+|---|---|---|
+| STRING | Used for STRING, CHAR and VARCHAR data. |
+
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-string-data-type.md:16]: @pokeeffe-molecula I'd like your thoughts on putting `cachetype lru size 1000` somewhere in here.

File: sql-preview/data-types/_posts/2022-11-16-cloud-string-data-type.md [hunk 17]
```diff
@@ -0,0 +1,25 @@
+---
+title: STRING data type
+---
+
+## Syntax
+
+```
+STRING...
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-string-data-type.md:8]: three dots?

might need to add something on `cachetype lru size 1000` somewhere in here. @pokeeffe-molecula

File: sql-preview/data-types/_posts/2022-11-16-cloud-stringset-data-type.md [hunk 18]
```diff
@@ -0,0 +1,34 @@
+---
+title: STRINGSET data type
+---
+
+## Syntax
+
+```
+STRINGSET...
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-stringset-data-type.md:8]: three dots? Might need to add `cachetype lru size 1000` somewhere in here @pokeeffe-molecula

File: sql-preview/data-types/_posts/2022-11-16-cloud-timequantum-ttl-constraint.md [hunk 19]
```diff
@@ -0,0 +1,112 @@
+---
+title: timeQuantum and TTL (Time To Live) Constraints
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-timequantum-ttl-constraint.md:2]: Is this page meant to be here. I think we need a reference page for how to apply these constraints here that points to further information on the existing TTL and time quantum pages (/concepts/time-quantums)

File: sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+```
+TIMESTAMP [TIMEUNIT] [EPOCH]
+```
+
+## Arguments
+
+| Argument | Description | Further information |
+|---|---|---|
+| TIMESTAMP | Time and date data type used for time series analysis | [Time stamp](https://en.wikipedia.org/wiki/Timestamp) |
+| TIMEUNIT | Constraint for time and date that defaults to second `s` | See [TIMEUNIT values](#timeunit-values) |
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md:16]: I think we should update the description to

> The time unit in which to store a timestamp that defaults to second `s`

Would like @pokeeffe-molecula take and to make sure that it is indeed about the unit we store things at.

File: sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+| EPOCH | The epoch which timestamps should be relative to. This is represented as a RFC339 time stamp string. | See [Epoch values](#epoch-values) |
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md:17]: @pokeeffe-molecula to confirm, EPOCH in SQL has nothing to do with ingest? This is simply how time is stored as an an int? Or am I off, and this is only specified if you ingest timestamps as integers?

File: sql-preview/data-types/_posts/2022-11-16-data-types-home.md [hunk 22]
```diff
@@ -0,0 +1,55 @@
+---
+title: Cloud table data types and constraints
+---
+
+Data types and constraints are used to define table columns when creating tables via the Cloud UI or the API.
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-data-types-home.md:5]: > Data types and constraints are used to define table columns when creating tables via SQL

File: sql-preview/data-types/_posts/2022-11-16-data-types-home.md [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+
+NOTE: For ease of use, Constraint information is included with the relevant Data type reference.
+
+## Before you begin
+
+* [Create a Cloud database using the UI](/cloud/cloud-databases/cloud-db-create), OR
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2022-11-16-data-types-home.md:11]: If we put these here, we'd want the equivalent for community edition as well

File: sql-preview/data-types/_posts/2023-01-01-cloud-time-quantum-constraint.md [hunk 24]
```diff
@@ -0,0 +1,7 @@
+---
+title: TIMEQUANTUM constraint
```
Comment (gthrone25) [inline sql-preview/data-types/_posts/2023-01-01-cloud-time-quantum-constraint.md:2]: Why are we adding this page? I think this is maybe a mistake from an old branch and already have a time quantume page?

**Review by gthrone25 — COMMENTED** [review]
Hey Lisa I tried to look at everything, but 60 files is a lot. As this is a PR about SQL and data types, many of the other file changes and tweaks (ingest, community, issues, etc.) ideally should have their own PRs and could go to prod without waiting on the reviews for sql and the rest.

There are a significant amount of bugs here (duped pages, missing nav, etc.) that I've tried to catch in the individual pages that need to be addressed. In general, I think this is a result of not starting with the fresh copy of the master branch (or copying all the files over from an old branch).

Really good start on the data type stuff! I've made comments throughout. One item i noticed is that all the pages are labeled cloud-* for the most part, but this applies to community as well, so we should remove cloud from the page names. I'm tagging in @pokeeffe-molecula to review and look for:

- [ ] Build the branch locally (see README for instructions)
- [ ] Content of new pages is accurate and clear (i.e. things are described correctly and are understandable)
- [ ] Content of new pages contains all information from removed pages (if applicable)
- [ ] Content of pages is comprehensive (i.e. all relevant content and features are documented and nothing is missing)
- [ ] Any other errors found (grammatical, spelling, consistent tense, etc.)

File: Gemfile [hunk 25]
```diff
@@ -41,3 +41,7 @@ gem 'jekyll-default-layout'
 #
 # For more info see: http://jekyll.github.io/github-metadata/
 gem "jekyll-github-metadata"
+
+# addon for htmlproofer dead link checker
+gem 'rake'
```
[the remark this answers is not in the document]
Comment (ljsinclair) [reply Gemfile:46]: rake is for the htmlproofer dead link checker which was an alternate checker I had to setup yesterday because of the problems with getting the site to build. It works slightly differently to the python script in that you build the site and it checks the generated HTML files.

File: sql-preview/data-types/_posts/2023-01-01-cloud-time-quantum-constraint.md (hunk 24 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2023-01-01-cloud-time-quantum-constraint.md:2]: Future dated page to capture some information that may be useful later.

File: sql-preview/data-types/_posts/2022-11-16-data-types-home.md (hunk 23 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-11-16-data-types-home.md:11]: Can you create a Community db with the UI? I didn't think that was possible?

File: sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md (hunk 4 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md:8]: 3 dots added because I dont know if there are any arguments
I can substitute with a comment #are there any arguments?

File: _includes/docs/community/fbcom-install-linux-source.md [hunk 26]
```diff
@@ -0,0 +1,25 @@
+### Step 1: Download FeatureBase
```
[the remark this answers is not in the document]
Comment (ljsinclair) [reply _includes/docs/community/fbcom-install-linux-source.md:1]: DOCS-101 was branched before we did all the work last week. It's got the issues from back then.
Once the changes have been made to the data-types files, I'll create a brand new branch, add them in and then do the broken link checks, merge checks et al.

File: sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md (hunk 5 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md:24]: Doh!

File: sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md (hunk 6 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-01-01-cloud-bool-data-type.md:28]: Ideally there'd be examples for all the data types and constraints. I'll remove this for now though

File: sql-preview/data-types/_posts/2022-11-16-cloud-idset-data-type.md (hunk 11 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-11-16-cloud-idset-data-type.md:15]: fixed

File: sql-preview/data-types/_posts/2022-11-16-cloud-timequantum-ttl-constraint.md (hunk 19 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-11-16-cloud-timequantum-ttl-constraint.md:2]: This is a hard one because there's *so much* information on timequantum and ttl.

My original thought was to put into an include file, but that got really unwieldly.

I'll take another look.

File: sql-preview/data-types/_posts/2022-11-16-data-types-home.md (hunk 3 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-11-16-data-types-home.md:50]: Yeah, cool.

File: sql-preview/data-types/_posts/2022-11-16-cloud-string-data-type.md (hunk 17 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-11-16-cloud-string-data-type.md]: removed 3 dots

File: sql-preview/data-types/_posts/2022-11-16-cloud-stringset-data-type.md (hunk 18 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-11-16-cloud-stringset-data-type.md]: removed 3 dots

File: sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md (hunk 20 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md]: updated description

File: sql-preview/data-types/_posts/2023-01-01-cloud-time-quantum-constraint.md (hunk 24 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2023-01-01-cloud-time-quantum-constraint.md]: deleted the page because the info is already in the standalone timequantum,ttl page

**Comment by ljsinclair** [comment]
Changes made:
* All changes from @gthrone25 
* filename changes because not cloud
* awaiting feedback from @pokeeffe-molecula 
Todo
* work out how to present timeQuantum and ttl info in a nicer way

**Comment by ljsinclair** [comment]
Changes made exclusively to the /sql-preview/data-types content. Once approved these files will be moved to a new branch.
/includes/sql-preview
/sql-preview/data-types

* All changes suggested by @greg are done
* filename changes because these are general-purpose data types
* setup include files for timequantum and ttl information, added to idset and stringset files
* tested for broken links (none in the /sql-preview/data-types folder, everything else isn't relevant)
* content builds and appears in the correct place.

@gthrone25 left 2 comments for @pokeeffe-molecula which are TODO

Once Pat has made comments, I'll create the new branch and copy the changed files across then create a new PR to merge

File: sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md (hunk 21 above)
Comment (pokeeffe-molecula) [reply sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md]: EPOCH is a column option, not necessarily a datatype thing, and there is a bit of conflation going on here.

The way to think about it is this:

- timestamp is a data type that represents a date/time value
- expressions/columns etc. can be of type timestamp
- epoch is an option that can be applied to a table column on table creation that governs how values (of type timestamp) are stored

And yes, nothing, directly to do with ingest.

File: sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md (hunk 20 above)
Comment (pokeeffe-molecula) [reply sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md]: TIMEUNIT is similar to EPOCH in that it is a column option used when creating/altering a table that governs how data (of type timestamp) is stored.

Comment (gthrone25) [reply sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md]: Updated value is accurate based on this comment. Resolving

File: sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md (hunk 21 above)
Comment (gthrone25) [reply sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md]: updating the description to reflect the answer above and committing up

Comment (gthrone25) [reply sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md]: actually @ljsinclair I don't have access to your branch and can't push. Can you update the description to:

> The epoch which timestamps will be stored relative to. This is represented as a RFC339 time stamp string.

**Comment by gthrone25** [comment]
@ljsinclair Besides the pat comments, the only thing I see that's kind of confusing is the multiple time quantum & ttl pages. If you can tell me structurally what you're trying to do, I can help write it, but today it's mixed with community stuff and repeated on some pages. Let's aim to get this in tomorrow!

**Comment by ljsinclair** [comment]
IMPORTANT -- new files from this branch/PR have been added to https://github.com/molecula/documentation/pull/280.
One that's merged, this PR can be closed WITHOUT merging :)
Thank you for your patience, normal service will resume very soon...

File: sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md (hunk 21 above)
Comment (ljsinclair) [reply sql-preview/data-types/_posts/2022-11-16-cloud-timestamp-data-type.md]: Updated in new-datatypes PR

**Comment by ljsinclair** [comment]
closing as this has its own PR now
